divi/src/main.cpp

```diff
@@ -1357,20 +1358,23 @@ void static InvalidBlockFound(CBlockIndex* pindex, const CValidationState& state
     }
 }
 
-void UpdateCoins(const CTransaction& tx, CCoinsViewCache& inputs, CTxUndo& txundo, int nHeight)
+void UpdateCoins(const CTransaction& tx, CCoinsViewCache& inputs, CTxUndo& txundo,
+                 const ActivationState& as, const int nHeight)
 {
     // mark inputs spent
     if (!tx.IsCoinBase() ) {
         txundo.vprevout.reserve(tx.vin.size());
-        BOOST_FOREACH (const CTxIn& txin, tx.vin) {
+        for (const auto& txin : tx.vin) {
             txundo.vprevout.push_back(CTxInUndo());
             bool ret = inputs.ModifyCoins(txin.prevout.hash)->Spend(txin.prevout.n, txundo.vprevout.back());
             assert(ret);
         }
     }
 
     // add outputs
-    inputs.ModifyCoins(tx.GetHash())->FromTx(tx, nHeight);
+    const uint256 outputHash = (as.IsActive(Fork::SegwitLight)
+                                  ? tx.GetBareTxid() : tx.GetHash());
+    inputs.ModifyCoins(outputHash)->FromTx(tx, nHeight);
```

Comment:
This doesn't seem very safe from an upgradability perspective. It seems like it is possible that an older transaction that was formerly referred to by legacy hash will now fail to be found by updated hash

Reply:
I'm not sure I understand your concern here.  A transaction will always be referred to based on the activation state at the block it was confirmed in.  There's no ambiguity there, and also e.g. the wallet will be able to properly handle spending any coins, created before or after the fork.

Perhaps a "weird" thing that could happen is if there is a reorg right around the fork, then a transaction might have been confirmed with e.g. old rules on one fork, and with new rules on another.  But even if that happens (which seems really unlikely in practice), that wouldn't cause any issues anywhere.